### ISSUE TITLE
extractors and their records should be identified by an extractor UUID and not name

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad-metalad#67]
feel free to close if that is the case already.
just came to mind so decided to check/ask:  if returned extracted metadata is returned as an associative array, keys should be some extractor's UUID and not a name to avoid collisions (different extensions might provide extractor with the same name), and extractor "entity" should just gain a name.

It also brings the question of what values to keep in `datalad.metadata.nativetype` -- names could also collide there rendering impossible to disambiguate.  But specifying UUIDs would be ugly/non-descriptive.  We could allow for names but that would imply that all extractors with that name would be used. And e.g. also allow for UUID specification or even `name:UUID` to disambiguate.

WDYT? or am I overthinking and current approach already would handle having multiple extractors with the same name smoothly?

**Comment by christian-monch** [comment]
Good point,  it was also raised in another discussion with @mih. Extractors should indeed, be identified by UUIDs in the future.

For new datasets and extractors that is simple, but it seems to me that moving to UUIDs needs some planning on the transition. We would most probably want to keep backward compatibility of existing extractors, extensions, and "datalad.metadata.nativetype"-entries in existing datasets.

**Comment by bpoldrack** [comment]
Wondered a while ago, what could be such an UUID. Thought of just content hashing the extractor, but then that would kinda limit extractors to be one file.

**Comment by christian-monch** [comment]
Added [backlog entry](https://github.com/datalad/datalad-metalad/projects/1#card-56672164) to [metadata-concept project](https://github.com/datalad/datalad-metalad/projects/1)

**Comment by christian-monch** [comment]
> Wondered a while ago, what could be such an UUID. Thought of just content hashing the extractor, but then that would kinda limit extractors to be one file.

Good questions. During the talk, we arrived at the conclusion that the extractor-author is responsible for assigning a UUID, e.g. in the code (currently by implementing the abstract get_id()-method).

**Comment by bpoldrack** [comment]
> During the talk, we arrived at the conclusion that the extractor-author is responsible for assigning a UUID, e.g. in the code (currently by implementing the abstract get_id()-method).

May be a combination in order to not allow extractor authors to mess things up: content hash of the file we know we are using + the their response to `get_id`? Although - also requires to work out how to interface arbitrary executables (in opposition to a python implementation).

**Comment by christian-monch** [comment]
I think content hashes might not be a good idea, because you could not update/bug-fix an extractor without changing its UUID.

**Comment by bpoldrack** [comment]
> I think content hashes might not be a good idea, because you could not update/bug-fix an extractor without changing its UUID.

Yes. Including a notion of version was the idea. Would generally see that as an advantage. It isn't the same extractor if it was updated.
However, the question will be how to address/configure extractors and whether that approach can cope with "versioned id".

**Comment by yarikoptic** [comment]
FWIW just 1c -- extractor UUID should be separate from its version.  UUID is just to give extractor a unique identity.  Generally should be completely random, but we might come up with some strategy to generate some deterministically for legacy ones for compatibility purposes even if at least for a while.  Fixed UUID prefix could be used to be possible to identify such legacy ones.
